### PR TITLE
Threaded safe creation

### PR DESCRIPTION
**Pull request by AssafMiron**

### CODE REVIEW

**Comment by infamousjoeg** [comment]
Will upgrade lab to v11.3 and test out.

**Comment by AssafMiron** [comment]
> Will upgrade lab to v11.3 and test out.

Hey @infamousjoeg ,
Any update?

**Comment by infamousjoeg** [comment]
> > Will upgrade lab to v11.3 and test out.
> 
> Hey @infamousjoeg ,
> Any update?

Sorry for the delay, testing this now...

File: Safe Management/Invoke-SafeManagement.ps1 [hunk 1]
```diff
@@ -0,0 +1,914 @@
...
+			Write-Host "Retrieving Safes..." -ForegroundColor Yellow
+			
+			try{
+				If (![string]::IsNullOrEmpty($SafeName))
+				{
+					Get-Safe -SafeName $SafeName
+				}
+				else
+				{
+					Get-Safes
```
Comment (infamousjoeg) [inline Safe Management/Invoke-SafeManagement.ps1:752]:
```
PS /Users/joegarcia/go/src/github.com/cyberark/epv-api-scripts/Safe Management> ./Invoke-SafeManagement.ps1 -PVWAURL https://cyberark.joegarcia.dev/PasswordVault -List                                                                     Script Started                                                                                                                                                                                                                              Setting script to use TLS 1.2                                                                                                                                                                                                                                                                                                                                                                                                                                                           Safe Management
Enter your User name and Password
User: jgarcia
Password for user jgarcia: ************

Retrieving Safes...
Retrieving safes from the vault...
There was an error retrieving the safes from the Vault. The error was:
Write-Error: /Users/joegarcia/go/src/github.com/cyberark/epv-api-scripts/Safe Management/Invoke-SafeManagement.ps1:752
Line |
 752 |                      Get-Safes
     |                      ~~~~~~~~~
     | Exception of type 'Microsoft.PowerShell.Commands.WriteErrorException' was thrown.

Logoff Session...
Script ended
```

File: Safe Management/Invoke-SafeManagement.ps1 [hunk 2]
```diff
@@ -0,0 +1,914 @@
...
+					}
+					if([string]::IsNullOrEmpty($NumVersionRetention))
+					{
+						$parameters.PSObject.Properties.Remove('numVersionRetention')
+					}
+					If($Add)
+					{
+						# Create one Safe
+						Write-Host "Adding the safe $SafeName..." -ForegroundColor Yellow
+						Create-Safe @parameters
```
Comment (infamousjoeg) [inline Safe Management/Invoke-SafeManagement.ps1:809]:
```
PS /Users/joegarcia/go/src/github.com/cyberark/epv-api-scripts/Safe Management> ./Invoke-SafeManagement.ps1 -PVWAURL https://cyberark.joegarcia.dev/PasswordVault -Add
Script Started                                                                                                                                                                                                                              Setting script to use TLS 1.2                                                                                                                                                                                                                                                                                                                                                                                                                                                           Safe Management
Enter your User name and Password
User: jgarcia
Password for user jgarcia: ************

Adding the safe ...

cmdlet Create-Safe at command pipeline position 1
Supply values for the following parameters:
safename: Test-Safe123
Adding the safe Test-Safe123 to the Vault...
Error adding Test-Safe123 to the Vault. The error was:
Write-Error: /Users/joegarcia/go/src/github.com/cyberark/epv-api-scripts/Safe Management/Invoke-SafeManagement.ps1:809
Line |
 809 |                          Create-Safe @parameters
     |                          ~~~~~~~~~~~~~~~~~~~~~~~
     | Exception of type 'Microsoft.PowerShell.Commands.WriteErrorException' was thrown.

Logoff Session...
Script ended
```

**Comment by infamousjoeg** [comment]
@AssafMiron I'm testing on PowerShell Core and not experiencing behavior I would expect out of this script.  What PS Version did you happen to test this on yourself during development?

**Comment by AssafMiron** [comment]
@infamousjoeg I'm using Powershell 5 for this
From the errors you encountered these are even without the Jobs and threaded mode so I guess the error is around the Logon Header
I changed it recently from Global to non-global and did not have time to check it yet, that might caused the issues
I added a way now to send the Logon header to all methods so that might solve things here
try running the script with -Debug -Verbose, this should show the REST Error and then we can verify it

Thanks,
Assaf

**Comment by bab29** [comment]
@AssafMiron 

This is almost a year old, do we need it or can I go ahead and close it?

**Comment by AssafMiron** [comment]
Think this can be closed
I think there is also anPR waiting for it, not sure if it was fully tested
or not

**Comment by bab29** [comment]
I will close this and if needed we can create a new one.

**Pull request closed without merging**